File: 2003-smallest-missing-genetic-value-in-each-subtree/2003-smallest-missing-genetic-value-in-each-subtree.js
/**
 * @param {number[]} parents
 * @param {number[]} nums
 * @return {number[]}
 */
var smallestMissingValueSubtree = function(parents, nums) {
    
    
    const tree = buildTree(parents)
    let currNode = nums.indexOf(1)
    let res = Array(nums.length).fill(1)
    let currSmall = 1
    let visited = Array(Math.max(...nums) * 2).fill(false) 
    while(currNode !== -1){
        dfs(currNode, visited, tree, nums)
        while(visited[currSmall])
            currSmall++
        res[currNode] = currSmall
        currNode = parents[currNode]
        
    }
    return res
};



function buildTree(parents){
    let tree = {}
    for(let [node, parent] of Object.entries(parents)){
        tree[parent] = tree[parent] || []
        tree[parent].push(node)
    }
    return tree
}


function dfs(node, visited, tree, nums){
        if(visited[nums[node]])return
        visited[nums[node]] = true
    for(let child of tree[node] || [])
        {
            dfs(child, visited, tree, nums)
        }
}

File: 2003-smallest-missing-genetic-value-in-each-subtree/2003-smallest-missing-genetic-value-in-each-subtree.py
class Solution:
    def smallestMissingValueSubtree(self, parents: List[int], nums: List[int]) -> List[int]:
        tree = defaultdict(list)
        for node, parent in enumerate(parents):
            tree[parent].append(node)
        seen = set()
        def dfs(node):
            if nums[node] in seen:
                return
            seen.add(nums[node])
            for nei in tree[node]:
                dfs(nei)
        if 1 not in nums:
            return [1] * len(parents)
        res = [1] * len(parents)
        node1 = nums.index(1)
        curr = 1

        while node1 != -1:
            dfs(node1)
            while curr in seen:
                curr += 1
            res[node1] = curr
            node1 = parents[node1]
        return res
        